modules/navier_stokes/include/base/INSFVMomentumResidualObject.h

```diff
@@ -0,0 +1,68 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "InputParameters.h"
+#include "INSFVRhieChowInterpolator.h"
+
+class MooseObject;
+class FaceInfo;
+
+/**
+ * All objects that contribute to pressure-based (e.g. not density-based) Navier-Stokes momentum
+ * equation residuals should inherit from this interface class. This holds true for INSFV, PINSFV,
+ * and WCNSFV objects (but not CNSFV or PCNSFV). This interface class introduces virtual methods
+ * that are used to gather data for Rhie-Chow interpolation and for interpolating and reconstructing
+ * body forces
+ */
+class INSFVMomentumResidualObject
+{
+public:
+  static InputParameters validParams();
+
+  /**
+   * @param obj the residual object that inherits from this interface
+   */
+  template <typename T>
+  INSFVMomentumResidualObject(T & obj);
+
+  /**
+   * Should be a non-empty implementation if the residual object is a \p FVElementalKernel and is
+   * not a pressure gradient term, e.g. this should be a non-empty implementation if the residual
+   * object introduces body forces. This will contribute to 'B' data structures where 'B' represents
+   * body forces in Moukalled notation
+   */
+  virtual void gatherRCData(const Elem & elem) = 0;
+
+  /**
+   * Should be a non-empty implementation if the residual object is a \p FVFluxKernel. This will
+   * contribute to 'a' data structures where 'a' represents on-diagonal momentum equation
+   * coefficients in Moukalled notation
+   */
+  virtual void gatherRCData(const FaceInfo & fi) = 0;
+
+  virtual ~INSFVMomentumResidualObject() = default;
+
+protected:
+  /// The Rhie-Chow user object that holds the 'a' and 'B' data structures which we will be feeding
```

Comment:
You should look at the 'Purge interpolation-reconstructions` commit ... I now only have the interpolator hold 'a' data and I do not conceptually make any distinction between 'a' vs. 'f'. E.g. if there is an on-diagonal coefficient, it goes into 'a'. I don't care if it's friction, viscosity, advection, or a whoosie-whatsie :smile: Interpolation of 'b' data is removed. It honestly never seemed to make much difference to me in the solution results nor in convergence (well unless you wrongfully tossed friction in there!). And I see no reason why a volumetric body force should be treated any differently than a prescribed flux boundary condition. In my mind the influences of body forces are held in the average velocity implicitly.

Reply:
Well actually interpolating 'b' did make a difference in solution results ... it smeared the body force and made it less accurate in truth. And if you really wanted body force smearing to help you with convergence, then you'd likely need more than one smoothing layer ... and then you'd need more ghosting. Anyway Moukalled creating a 'b' and a 'B' and then treating them differently just seems like heuristic wish-wash to me. I don't like it. I *really* like what we now have going on with 'a' including friction ... per the most recent vtb PBFHR results I feel as if we've added solution stability and maintained good results

Reply:
Ah ok I was going by files first, not by commits, I ll see this soon then. 
Good to know though, means I have to re-generate the steady results for the MSFR so we can have a good restart for the MSFR transient. 
I agree that the multiplication of smoothing layers was going to be an issue with precision. I was not opposed to the B_1 - B_3 term. It seems heuristic but it's justified in Moukalled for two reasons: better robustness for free surface flows (with big variation of body forces at the surface) and to get equilibrium between pressure gradient and body forces, which ought to share the same stencil as he derives

Reply:
Might be worth running a still case with Boussinesq and seeing if we have issues. i suspect not since you were getting good results for a variety of Ra before this PR already

Reply:
What do you mean by a still case?

Reply:
water not moving. I guess if there's temperature gradients, water is always moving?

Reply:
If water is not moving, why are we trying to simulate it as a fluid?

Reply:
Oh, sorry I just read the rest of the comments. Disregard my question.

Reply:
I've created a `keep-b-in-rc` branch if you want to continue to investigate. I'm happy with the much less complex user object and seemingly better solve convergence without B in RC. But of course I always maintain an open mind

Reply:
We could also keep the code around but make the default `standard_body_forces = true`

Reply:
I'm actually looking at 'still' results for the cask and they look bad with and without the body force treatment

Reply:
Is this natural convection? If so, what is the Rayleigh number and how much does the density vary? Should it be using weakly compressible if it isn't already? What is the `advected_interp_method`? Can you grant me access if it's not public?

Reply:
see slack